### ISSUE TITLE
Delete example tiny-cluster is not deleting the resources created

### ISSUE THREAD

**Issue by ankitmani2004** [druid-io/druid-operator#182]
Hi,

I have deployed the example tiny-cluster and trying to re-deploy with some changes. When tried deleting the already deployed tiny-cluster, Services and tiny-cluster is not getting deleted(manually deleted STS).

Thanks.

**Comment by AdheipSingh** [comment]
can you send some info about events from operator ```kubectl describe druid -n <namespace>``` and some logs from operator. 

Operator shall trigger finalizer so sts, pvc are definelty to get cleanedup. How are deleting the cluster ?

**Comment by ankitmani2004** [comment]
Hi,

I am using kubectl delete -f tiny-cluster.yaml command to delete the cluster. Here its o/p of describe:
`Name:         tiny-cluster
Namespace:    pmfns01
Labels:       <none>
Annotations:  <none>
API Version:  druid.apache.org/v1alpha1
Kind:         Druid
Metadata:
  Creation Timestamp:             2021-05-11T13:58:29Z
  Deletion Grace Period Seconds:  0
  Deletion Timestamp:             2021-05-11T14:14:59Z
  Finalizers:
    deletepvc.finalizers.druid.apache.org
  Generation:  4
  Managed Fields:
    API Version:  druid.apache.org/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
      f:spec:
        f:nodes:
          f:brokers:
            f:resources:
          f:coordinators:
            f:resources:
          f:historicals:
            f:resources:
          f:routers:
            f:resources:
      f:status:
    Manager:      manager
    Operation:    Update
    Time:         2021-05-11T13:58:29Z
    API Version:  druid.apache.org/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:common.runtime.properties:
        f:commonConfigMountPath:
        f:env:
        f:image:
        f:jvm.options:
        f:log4j.config:
        f:nodes:
          .:
          f:brokers:
            .:
            f:druid.port:
            f:extra.jvm.options:
            f:nodeConfigMountPath:
            f:nodeType:
            f:replicas:
            f:runtime.properties:
          f:coordinators:
            .:
            f:druid.port:
            f:extra.jvm.options:
            f:nodeConfigMountPath:
            f:nodeType:
            f:replicas:
            f:runtime.properties:
          f:historicals:
            .:
            f:druid.port:
            f:extra.jvm.options:
            f:nodeConfigMountPath:
            f:nodeType:
            f:replicas:
            f:runtime.properties:
          f:routers:
            .:
            f:druid.port:
            f:extra.jvm.options:
            f:nodeConfigMountPath:
            f:nodeType:
            f:replicas:
            f:runtime.properties:
        f:podAnnotations:
          .:
          f:dummykey:
        f:podLabels:
          .:
          f:environment:
          f:release:
        f:readinessProbe:
          .:
          f:httpGet:
            .:
            f:path:
            f:port:
        f:securityContext:
          .:
          f:fsGroup:
          f:runAsGroup:
          f:runAsUser:
        f:services:
        f:startScript:
        f:volumeMounts:
        f:volumes:
    Manager:         kubectl-client-side-apply
    Operation:       Update
    Time:            2021-05-11T14:26:15Z
  Resource Version:  17811345059
  Self Link:         /apis/druid.apache.org/v1alpha1/namespaces/pmfns01/druids/tiny-cluster
  UID:               dcc9db60-72da-4b9c-9c11-b21d0c40510b
Spec:
  common.runtime.properties:  
# Zookeeper
druid.zk.service.host=tiny-cluster-zk-0.tiny-cluster-zk
druid.zk.paths.base=/druid
druid.zk.service.compress=false

# Metadata Store
druid.metadata.storage.type=derby
druid.metadata.storage.connector.connectURI=jdbc:derby://localhost:1527/druid/data/derbydb/metadata.db;create=true
druid.metadata.storage.connector.host=localhost
druid.metadata.storage.connector.port=1527
druid.metadata.storage.connector.createTables=true

# Deep Storage
druid.storage.type=local
druid.storage.storageDirectory=/druid/deepstorage

#
# Extensions
#
druid.extensions.loadList=["druid-kafka-indexing-service"]

#
# Service discovery
#
druid.selectors.indexing.serviceName=druid/overlord
druid.selectors.coordinator.serviceName=druid/coordinator

druid.indexer.logs.type=file
druid.indexer.logs.directory=/druid/data/indexing-logs
druid.lookup.enableLookupSyncOnStartup=false

  Common Config Mount Path:  /opt/druid/conf/druid/cluster/_common
  Env:
    Name:  POD_NAME
    Value From:
      Field Ref:
        Field Path:  metadata.name
    Name:            POD_NAMESPACE
    Value From:
      Field Ref:
        Field Path:  metadata.namespace
  Image:             apache/druid:0.20.0
  jvm.options:       -server
-XX:MaxDirectMemorySize=10240g
-Duser.timezone=UTC
-Dfile.encoding=UTF-8
-Dlog4j.debug
-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager
-Djava.io.tmpdir=/druid/data
  log4j.config:  <?xml version="1.0" encoding="UTF-8" ?>
<Configuration status="WARN">
    <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{ISO8601} %p [%t] %c - %m%n"/>
        </Console>
    </Appenders>
    <Loggers>
        <Root level="info">
            <AppenderRef ref="Console"/>
        </Root>
    </Loggers>
</Configuration>
  Nodes:
    Brokers:
      druid.port:         8088
      extra.jvm.options:  -Xmx512M
-Xms512M
      Node Config Mount Path:  /opt/druid/conf/druid/cluster/query/broker
      Node Type:               broker
      Replicas:                1
      Resources:
      runtime.properties:  druid.service=druid/broker

# HTTP server threads
druid.broker.http.numConnections=5
druid.server.http.numThreads=10

# Processing threads and buffers
druid.processing.buffer.sizeBytes=1
druid.processing.numMergeBuffers=1
druid.processing.numThreads=1
druid.sql.enable=true

    Coordinators:
      druid.port:         8088
      extra.jvm.options:  -Xmx512M
-Xms512M
      Node Config Mount Path:  /opt/druid/conf/druid/cluster/master/coordinator-overlord
      Node Type:               coordinator
      Replicas:                1
      Resources:
      runtime.properties:  druid.service=druid/coordinator

# HTTP server threads
druid.coordinator.startDelay=PT30S
druid.coordinator.period=PT30S

# Configure this coordinator to also run as Overlord
druid.coordinator.asOverlord.enabled=true
druid.coordinator.asOverlord.overlordService=druid/overlord
druid.indexer.queue.startDelay=PT30S
druid.indexer.runner.type=local

    Historicals:
      druid.port:         8088
      extra.jvm.options:  -Xmx512M
-Xms512M
  
      Node Config Mount Path:  /opt/druid/conf/druid/cluster/data/historical
      Node Type:               historical
      Replicas:                1
      Resources:
      runtime.properties:  druid.service=druid/historical
druid.server.http.numThreads=5
druid.processing.buffer.sizeBytes=536870912
druid.processing.numMergeBuffers=1
druid.processing.numThreads=1
# Segment storage
druid.segmentCache.locations=[{\"path\":\"/druid/data/segments\",\"maxSize\":10737418240}]
druid.server.maxSize=10737418240

    Routers:
      druid.port:         8088
      extra.jvm.options:  -Xmx512M
-Xms512M
      Node Config Mount Path:  /opt/druid/conf/druid/cluster/query/router
      Node Type:               router
      Replicas:                1
      Resources:
      runtime.properties:  druid.service=druid/router

# HTTP proxy
druid.router.http.numConnections=10
druid.router.http.readTimeout=PT5M
druid.router.http.numMaxThreads=10
druid.server.http.numThreads=10

# Service discovery
druid.router.defaultBrokerServiceName=druid/broker
druid.router.coordinatorServiceName=druid/coordinator

# Management proxy to coordinator / overlord: required for unified web console.
druid.router.managementProxy.enabled=true       

  Pod Annotations:
    Dummykey:  dummyval
  Pod Labels:
    Environment:  stage
    Release:      alpha
  Readiness Probe:
    Http Get:
      Path:  /status/health
      Port:  8088
  Security Context:
    Fs Group:      1000
    Run As Group:  1000
    Run As User:   1000
  Services:
    Spec:
      Cluster IP:  None
      Type:        ClusterIP
  Start Script:    /druid.sh
  Volume Mounts:
    Mount Path:  /druid/data
    Name:        data-volume
    Mount Path:  /druid/deepstorage
    Name:        deepstorage-volume
  Volumes:
    Empty Dir:
    Name:  data-volume
    Empty Dir:
    Name:  deepstorage-volume
Events:    <none>`

**Comment by ankitmani2004** [comment]
some logs 
`kubectl logs -f druid-operator-bbbb47c7-ftshs --namespace pmfns01
I0512 07:49:39.952967   99792 request.go:645] Throttling request took 1.193699745s, request: GET:https://<url>:443/apis/jaegertracing.io/v1?timeout=32s
I0511 13:55:05.013993       1 request.go:621] Throttling request took 1.046738386s, request: GET:https://192.168.0.1:443/apis/autoscaling/v2beta1?timeout=32s
2021-05-11T13:55:08.169Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": ":8080"}
2021-05-11T13:55:08.171Z	INFO	setup	starting manager
2021-05-11T13:55:08.171Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2021-05-11T13:55:08.171Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "druid", "source": "kind source: /, Kind="}
2021-05-11T13:55:08.271Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "druid"}
2021-05-11T13:55:08.272Z	INFO	controller-runtime.controller	Starting workers	{"controller": "druid", "worker count": 1}
2021-05-11T13:58:29.909Z	INFO	druid_operator_handler	Created [:tiny-cluster-druid-common-config].	{"Object": "{\"metadata\":{\"name\":\"tiny-cluster-druid-common-config\",\"namespace\":\"pmfns01\",\"selfLink\":\"/api/v1/namespaces/pmfns01/configmaps/tiny-cluster-druid-common-config\",\"uid\":\"de28855c-3ae4-41f6-be9b-cbc90800a54b\",\"resourceVersion\":\"17810083075\",\"creationTimestamp\":\"2021-05-11T13:58:29Z\",\"labels\":{\"app\":\"druid\",\"druid_cr\":\"tiny-cluster\"},\"annotations\":{\"druidOpResourceHash\":\"z9fvVjefK7TekLTTR3x1r0CF9n8=\"},\"ownerReferences\":[{\"apiVersion\":\"druid.apache.org/v1alpha1\",\"kind\":\"Druid\",\"name\":\"tiny-cluster\",\"uid\":\"dcc9db60-72da-4b9c-9c11-b21d0c40510b\",\"controller\":true}],\"managedFields\":[{\"manager\":\"manager\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-05-11T13:58:29Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:data\":{\".\":{},\"f:common.runtime.properties\":{}},\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:druidOpResourceHash\":{}},\"f:labels\":{\".\":{},\"f:app\":{},\"f:druid_cr\":{}},\"f:ownerReferences\":{\".\":{},\"k:{\\\"uid\\\":\\\"dcc9db60-72da-4b9c-9c11-b21d0c40510b\\\"}\":{\".\":{},\"f:apiVersion\":{},\"f:controller\":{},\"f:kind\":{},\"f:name\":{},\"f:uid\":{}}}}}}]},\"data\":{\"common.runtime.properties\":\"\\n# Zookeeper\\ndruid.zk.service.host=tiny-cluster-zk-0.tiny-cluster-zk\\ndruid.zk.paths.base=/druid\\ndruid.zk.service.compress=false\\n\\n# Metadata Store\\ndruid.metadata.storage.type=derby\\ndruid.metadata.storage.connector.connectURI=jdbc:derby://localhost:1527/druid/data/derbydb/metadata.db;create=true\\ndruid.metadata.storage.connector.host=localhost\\ndruid.metadata.storage.connector.port=1527\\ndruid.metadata.storage.connector.createTables=true\\n\\n# Deep Storage\\ndruid.storage.type=local\\ndruid.storage.storageDirectory=/druid/deepstorage\\n\\n#\\n# Extensions\\n#\\ndruid.extensions.loadList=[\\\"druid-kafka-indexing-service\\\"]\\n\\n#\\n# Service discovery\\n#\\ndruid.selectors.indexing.serviceName=druid/overlord\\ndruid.selectors.coordinator.serviceName=druid/coordinator\\n\\ndruid.indexer.logs.type=file\\ndruid.indexer.logs.directory=/druid/data/indexing-logs\\ndruid.lookup.enableLookupSyncOnStartup=false\\n\"}}", "name": "tiny-cluster", "namespace": "pmfns01"}
`
and also `E0511 13:58:30.608038       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to list *v1beta1.Ingress: ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:pmfns01:druid-operator" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "pmfns01"
`

**Comment by layoaster** [comment]
I'm hitting the same issue, in my case I'm using the helm chart to deploy the operator. My findings: 

1. For namespaced-scoped deployments you must include `watchNamespace: true` in `values.yaml`  to get the `Role` and `RoleBinding` objects generated.

2. The Role/RoleBinding are generated with the namespace of the Helm release so if you plan to deploy the Druid cluster in the `default` namespace (or another namespace != Druid-operator Namespace) the service account will still have no permissions on that namespace

**Comment by AdheipSingh** [comment]
@layoaster 

I get your issue, ideally the  ```watchNamespace: true``` should not be added, and the role binding should be generated on just ```env.WATCH_NAMESPACE```

The issue lies here https://github.com/druid-io/druid-operator/blob/master/chart/templates/role.yaml#L3 , we can change this too ```{{- if .Values.env.WATCH_NAMESPACE }}```  , in case ```env.WATCH_NAMESPACE``` is kept to ```""``` , cluster scoped rbac will be create , else namespace scoped.

**Comment by layoaster** [comment]
@AdheipSingh that's is what I did first but wasn't sure it was the right way to do it, so I decided in favor of the `watchNamespace: true` which felt less intrusive

**Comment by layoaster** [comment]
also the `Role definition` is missing the `networking.k8s.io` API Group:

```yaml
- apiGroups:
    - extensions
    - networking.k8s.io
  resources:
    - ingresses
  verbs:
    - '*'
```

**Comment by ankitmani2004** [comment]
I did not deploy using helm charts, can you suggest @AdheipSingh . Thanks

**Comment by AdheipSingh** [comment]
@ankitmani2004 the logs arnt enought to find the root cause. Can you issue a delete command and the send the latest logs. Any deletion error should be catched their.

**Comment by AdheipSingh** [comment]
@layoaster do you mind sending a PR for the above fixes ?

**Comment by ankitmani2004** [comment]
I just issued:
kubectl delete -f examples/tiny-cluster.yaml  --grace-period=0 --force
druid.druid.apache.org "tiny-cluster" force deleted

and the screen is stuck here. In other tab I ran below:

kubectl logs -f druid-operator-bbbb47c7-ftshs --namespace pmfns01

and the log keeps repeating the below error:

E0512 18:28:05.224034       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to list *v1beta1.Ingress: ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:pmfns01:druid-operator" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "pmfns01"
E0512 18:28:57.551093       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to list *v1beta1.Ingress: ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:pmfns01:druid-operator" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "pmfns01"

**Comment by layoaster** [comment]
> @layoaster do you mind sending a PR for the above fixes ?

Sure, I'll do it

**Comment by layoaster** [comment]
> I just issued:
> kubectl delete -f examples/tiny-cluster.yaml --grace-period=0 --force
> druid.druid.apache.org "tiny-cluster" force deleted
> 
> and the screen is stuck here. In other tab I ran below:
> 
> kubectl logs -f druid-operator-bbbb47c7-ftshs --namespace pmfns01
> 
> and the log keeps repeating the below error:
> 
> E0512 18:28:05.224034 1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to list *v1beta1.Ingress: ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:pmfns01:druid-operator" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "pmfns01"
> E0512 18:28:57.551093 1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.2/tools/cache/reflector.go:125: Failed to list *v1beta1.Ingress: ingresses.networking.k8s.io is forbidden: User "system:serviceaccount:pmfns01:druid-operator" cannot list resource "ingresses" in API group "networking.k8s.io" in the namespace "pmfns01"

@ankitmani2004 I can confirm that the permissions issue is preventing the cluster (druid spec) from being removed.

The workaround is to modify the `Role`/`ClusterRole` to add the `networking.k8s.io` API Group. You should have something like this:

```yaml
- apiGroups:
    - extensions
    - networking.k8s.io
  resources:
    - ingresses
  verbs:
    - '*'
```
Last but not least: if you are using a namespace scoped deployment, you must make sure the role and its binding have the `namespace` properly set.

**Comment by ankitmani2004** [comment]
For me deleting the  "deletepvc.finalizers.druid.apache.org" line from tiny-cluster worked to delete the resources.

**Comment by AdheipSingh** [comment]
@ankitmani2004 can you please create a proper gist of your yaml. So i can re-run it. I have not been able to recreate the issue on tiny-cluster. The role of finalizer is too clean up sts + pvc when a deletion Timestamp is set on the CR.

**Comment by himanshug** [comment]
Fixed by #184
